cloudinit/sources/helpers/azure.py

```diff
@@ -177,6 +185,46 @@ def report_diagnostic_event(str):
     return evt
 
 
+def report_compressed_event(event_name, event_content):
+    """Report a compressed event"""
+    compressed_data = base64.encodebytes(zlib.compress(event_content))
+    event_data = {"encoding": "gz+b64",
+                  "data": compressed_data.decode('ascii')}
+    evt = events.ReportingEvent(
+        COMPRESSED_EVENT_TYPE, event_name,
+        json.dumps(event_data),
+        events.DEFAULT_EVENT_ORIGIN)
+    events.report_event(evt, excluded_handlers={"logging"})
+
+    # return the event for unit testing purpose
+    return evt
+
+
+@azure_ds_telemetry_reporter
+def push_log_to_kvp(file_name=CFG_BUILTIN['def_log_file']):
+    """Push a portion of cloud-init.log file or the whole file to KVP
+    based on the file size.
+    When called more than once, the function continues pushing the log file
+    from where it left off last time it was called."""
+    global last_log_byte_pushed_to_kvp_index
```

Comment:
Since cloud-init spins up a new process across different boot stages. I *think* this means that you will start at byte zero if push_log_to_kvp is called in init-local stage and init(network active) stage.
It also means that each time you reboot you will be sending that entire log compressed again to hyperv's kvp if we call down this path.  I'm checking to see if we can trigger this call across each reboot (it looks like anytime we'd call crawl-metadata on the azure datasource we would get down here and re-push the full cloud-init.log

Reply:
Confirmed, rebooting an azure instance you will always get a full compressed aggregated cloud-init.log message across reboots. So, the more times we reboot, the bigger this message gets.
I added the byte offset to the push_log_to_kvp debug message
    LOG.debug("Dumping cloud-init.log file to KVP %d", last_log_byte_pushed_to_kvp_index)

which results in the following across multiple boots:
```
2020-08-18 18:59:41,564 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP
2020-08-18 18:59:41,564 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 54916 bytes of cloud-init.log file to KVP
2020-08-18 18:59:41,741 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP
2020-08-18 18:59:41,741 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 12612 bytes of cloud-init.log file to KVP
2020-08-18 19:02:49,612 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP
2020-08-18 19:02:49,612 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 207155 bytes of cloud-init.log file to KVP
2020-08-18 19:02:49,797 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP
2020-08-18 19:02:49,797 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 11951 bytes of cloud-init.log file to KVP
2020-08-18 19:04:29,861 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP 0
2020-08-18 19:04:29,862 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 330507 bytes of cloud-init.log file to KVP
2020-08-18 19:04:30,054 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP 330636
2020-08-18 19:04:30,054 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 11958 bytes of cloud-init.log file to KVP
2020-08-18 19:05:20,768 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP 0
2020-08-18 19:05:20,769 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 453650 bytes of cloud-init.log file to KVP
2020-08-18 19:05:20,966 - azure.py[DEBUG]: Dumping cloud-init.log file to KVP 453779
2020-08-18 19:05:20,966 - handlers.py[DEBUG]: diagnostic: diagnostic message: Dumping 11958 bytes of cloud-init.log file to KVP
```

Reply:
@blackboxsw this is a good observation. I think we want to avoid pushing it every boot and rather just on first boot. Perhaps it needs an is-new-instance check?

Reply:
@anhvoms @Moustafa-Moustafa I think that makes sense. if you only care about this full log on first boot, an is_new_instance check gating the call to push_log_to_kvp. I see that that function is called from both DataSourceAzure.activate() and also sources/helpers/azure.py send_ready_signal(). If you don't care to have this compressed log data on secondary boot, it'd be easy from the DataSourceAzure.activate() to avoid that with `if not self.check_instance_id():\n    push_log_to_kvp()`.

I'm not certain how you should perform that check in get_metadata_from_fabric() which calls send_ready_signal()

Reply:
KVPs are only collected from the platform during first boot (when VM reports ready, or fails to report ready after some X minutes) so anything after that don't really make a difference. The KVP daemon on the VM still processes any KVPs written, so we do want to limit writing unnecessary KVPs 

Only new instance should go through the path to send ready signal so I think that code path doesn't need a new instance check?

Reply:
I think that makes sense. Though I do see on Azure that, by default, _get_data  is performed on each boot of Azure vms (in case network config changed). So, as any calls to push_log_to_kvp are gated by a check_instance_id()  I think you should be safe.

```
2020-08-18 19:04:26,713 - __init__.py[DEBUG]: Update datasource metadata and network config due to events: System boot
2020-08-18 19:04:26,713 - handlers.py[DEBUG]: start: azure-ds/_get_data: _get_data
```

Reply:
Yep - I just deployed a VM and confirmed that we do go through that path and report ready on every boot. Gating the call to push_log_to_kvp by check_instance_id is the right approach I think